crates/common/types/transaction.rs

```diff
@@ -35,6 +37,116 @@ pub enum Transaction {
+    PrivilegedL2Transaction(PrivilegedL2Transaction),
+}
+
+impl TryInto<Transaction> for P2PTransaction {
+    type Error = String;
+
+    fn try_into(self) -> Result<Transaction, Self::Error> {
+        match self {
+            P2PTransaction::LegacyTransaction(itx) => Ok(Transaction::LegacyTransaction(itx)),
+            P2PTransaction::EIP2930Transaction(itx) => Ok(Transaction::EIP2930Transaction(itx)),
+            P2PTransaction::EIP1559Transaction(itx) => Ok(Transaction::EIP1559Transaction(itx)),
+            P2PTransaction::PrivilegedL2Transaction(itx) => {
+                Ok(Transaction::PrivilegedL2Transaction(itx))
+            }
+            _ => Err("Can't convert blob p2p transaction into regular transaction. Blob bundle would be lost.".to_string()),
+        }
+    }
+}
+
+impl RLPEncode for P2PTransaction {
+    fn encode(&self, buf: &mut dyn bytes::BufMut) {
+        match self {
+            P2PTransaction::LegacyTransaction(t) => t.encode(buf),
+            tx => Bytes::copy_from_slice(&tx.encode_canonical_to_vec()).encode(buf),
+        };
+    }
+}
+
+impl RLPDecode for P2PTransaction {
+    fn decode_unfinished(rlp: &[u8]) -> Result<(Self, &[u8]), RLPDecodeError> {
+        if is_encoded_as_bytes(rlp)? {
+            // Adjust the encoding to get the payload
+            let payload = get_rlp_bytes_item_payload(rlp);
+            let tx_type = payload.first().unwrap();
+            let tx_encoding = &payload[1..];
+            // Look at the first byte to check if it corresponds to a TransactionType
+            match *tx_type {
+                // Legacy
+                0x0 => LegacyTransaction::decode_unfinished(tx_encoding)
+                    .map(|(tx, rem)| (P2PTransaction::LegacyTransaction(tx), rem)), // TODO: check if this is a real case scenario
+                // EIP2930
+                0x1 => EIP2930Transaction::decode_unfinished(tx_encoding)
+                    .map(|(tx, rem)| (P2PTransaction::EIP2930Transaction(tx), rem)),
+                // EIP1559
+                0x2 => EIP1559Transaction::decode_unfinished(tx_encoding)
+                    .map(|(tx, rem)| (P2PTransaction::EIP1559Transaction(tx), rem)),
+                // EIP4844
+                0x3 => WrappedEIP4844Transaction::decode_unfinished(tx_encoding)
+                    .map(|(tx, rem)| (P2PTransaction::WrappedEIP4844Transaction(tx), rem)),
+                // PriviligedL2
+                0x7e => PrivilegedL2Transaction::decode_unfinished(tx_encoding)
+                    .map(|(tx, rem)| (P2PTransaction::PrivilegedL2Transaction(tx), rem)),
+                ty => Err(RLPDecodeError::Custom(format!(
+                    "Invalid transaction type: {ty}"
+                ))),
+            }
+        } else {
+            // LegacyTransaction
+            LegacyTransaction::decode_unfinished(rlp)
+                .map(|(tx, rem)| (P2PTransaction::LegacyTransaction(tx), rem))
+        }
+    }
+}
```

Comment:
Once again, this pattern rises, the distinction of encoding typed vs legacy transactions, could we leverage the fact that the transaction type already implements this?